homeassistant/components/template/vacuum.py

```diff
@@ -0,0 +1,364 @@
+                hass, device, friendly_name,
+                state_template, battery_level_template, fan_speed_template,
+                start_action, pause_action, stop_action, return_to_base_action,
+                clean_spot_action, locate_action, set_fan_speed_action,
+                fan_speed_list, entity_ids
+            )
+        )
+
+    async_add_entities(vacuums)
+
+
+class TemplateVacuum(StateVacuumDevice):
+    """A template vacuum component."""
+
+    def __init__(self, hass, device_id, friendly_name,
+                 state_template, battery_level_template, fan_speed_template,
+                 start_action, pause_action, stop_action,
+                 return_to_base_action, clean_spot_action, locate_action,
+                 set_fan_speed_action, fan_speed_list, entity_ids):
+        """Initialize the vacuum."""
+        self.hass = hass
+        self.entity_id = async_generate_entity_id(
+            ENTITY_ID_FORMAT, device_id, hass=hass)
+        self._name = friendly_name
+
+        self._template = state_template
+        self._battery_level_template = battery_level_template
+        self._fan_speed_template = fan_speed_template
+        self._supported_features = SUPPORT_START
+
+        self._start_script = Script(hass, start_action)
+
+        self._pause_script = None
+        if pause_action:
+            self._pause_script = Script(hass, pause_action)
+            self._supported_features |= SUPPORT_PAUSE
+
+        self._stop_script = None
+        if stop_action:
+            self._stop_script = Script(hass, stop_action)
+            self._supported_features |= SUPPORT_STOP
+
+        self._return_to_base_script = None
+        if return_to_base_action:
+            self._return_to_base_script = Script(hass, return_to_base_action)
+            self._supported_features |= SUPPORT_RETURN_HOME
+
+        self._clean_spot_script = None
+        if clean_spot_action:
+            self._clean_spot_script = Script(hass, clean_spot_action)
+            self._supported_features |= SUPPORT_CLEAN_SPOT
+
+        self._locate_script = None
+        if locate_action:
+            self._locate_script = Script(hass, locate_action)
+            self._supported_features |= SUPPORT_LOCATE
+
+        self._set_fan_speed_script = None
+        if set_fan_speed_action:
+            self._set_fan_speed_script = Script(hass, set_fan_speed_action)
+            self._supported_features |= SUPPORT_FAN_SPEED
+
+        self._state = None
+        self._battery_level = None
+        self._fan_speed = None
+
+        if self._template:
+            self._template.hass = self.hass
+            self._supported_features |= SUPPORT_STATE
+        if self._battery_level_template:
+            self._battery_level_template.hass = self.hass
+            self._supported_features |= SUPPORT_BATTERY
+        if self._fan_speed_template:
+            self._fan_speed_template.hass = self.hass
+            self._supported_features |= SUPPORT_FAN_SPEED
```

Comment:
Either a speed template or a speed script will make speed supported? Is this correct?

Reply:
It doesn't look correct. The fan speed method requires `self._set_fan_speed_script` to be truthy. So we should only support speed if there's a speed script.